### ISSUE TITLE
Remove Jython platform detection

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3725]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3725|
|Type|enhancement|
|Created|2009-03-30 22:53:12Z|
|Branch|https://github.com/twisted/twisted/tree/jython-platformdetection-remove-3725-2|

Recent versions of Jython are totally different from old versions of Jython which Twisted supported.  The current code is mostly just getting in the way of things.  We should drop it so that it's clear what we _should_ do in order to support the new version of Jython.

<details><summary>Searchable metadata</summary>

```
trac-id__3725 3725
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_jython_platformdetection_remove_3725_2 branches/jython-platformdetection-remove-3725-2
branch_author__cyli cyli
status__closed closed
resolution__fixed fixed
component__core core
keywords__jython jython
time__1238453592000000 1238453592000000
changetime__1274596444000000 1274596444000000
version__None None
owner__ 
cc__thijs cc__ltaylor.volks cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_1"></a>|@thijstriemstra commented|
|-|-|

Are you referring to the code in setup.py? Where can this Jython platform detection be found?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

I mean everything throughout Twisted which has a special case in order to support Jython.  So, for example:

  * twisted.spread.jelly, around line 121
  * twisted.internet._javaserialport
  * twisted.python.threadpool, around line 60

**Comment by twisted-trac** [comment]
|<img alt="ltaylor.volks's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|ltaylor.volks commented|
|-|-|

As of changeset:12375, there is no javareactor, so twisted.internet._javaserialport won't work at all and should be able to be safely removed, along with the import in twisted.internet.serialport (line 62):
```
elif os.name == 'java':
    from twisted.internet._javaserialport import SerialPort
```

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_4"></a>|@cyli commented|
|-|-|

(In [28688]) Branching to 'jython-platformdetection-remove-3725'

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_5"></a>|@cyli removed owner|
|-|-|

Removed relevant platform checks from spread/jelly.py, internet/serialport.py, and python/threadpool.py (also changed test_threadpool to test deprecation of ThreadSafeList).  And removed the _javaserialport.py file.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_7"></a>|@jesstess set owner to @cyli|
|-|-|

Thanks for working on this, cyli. The meat of the ticket work looks great. Some feedback:
 * I might reword the deprecation warning for ThreadSafeList to make it sound less like Twisted fosters dubious and broken code. :) Maybe something like "This code supported the now obsolete Jython 2.1 and is deprecated to make way for support for newer Jython versions". Maybe with less verbosity.
 * 3725.removal needs a final carriage return
 * test_threadpool.py and threadpool.py have trailing whitespace
 * does serialport.py really not have any test cases? I ask because it doesn't have a test case name at the top, but grepping around I didn't see anything. Too bad.
 * the for loop at the bottom of test_threadpool.py for setting skip uses DeprecationTestCase when the test class's name is ThreadSafeListDeprecationTestCase. This [doesn't appear to have actually broken anything](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/jython-platformdetection-remove-3725), so is the surrounding interface check useless?
 * pyflakes warnings: test_threadpool.py imports Version but doesn't use it. threadpool.py imports runtime but doesn't use it. serialport.py has a bunch of unused imports. jelly.py imports runtime but doesn't use it.
 * grepping through the source, I see a couple of other places where something about Java is special-cased, although I didn't check if they fall under the umbrella of this ticket: setup.py, twisted/python/runtime.py, and twisted/test/test_pb.py.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

> 3725.removal needs a final carriage return 

Probably a good idea, but the news generator won't actually care.  It reformats the text to fit in the news file anyway.

> does serialport.py really not have any test cases?

Sad and true.  But [#2462](https://github.com/twisted/twisted/issues/2462), [#3690](https://github.com/twisted/twisted/issues/3690), [#4249](https://github.com/twisted/twisted/issues/4249). :)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_9"></a>|@cyli commented|
|-|-|

>  * I might reword the deprecation warning for ThreadSafeList to make it sound less like Twisted fosters dubious and broken code. :) Maybe something like "This code supported the now obsolete Jython 2.1 and is deprecated to make way for support for newer Jython versions". Maybe with less verbosity.

Ah ok, I was given to understand that it never worked in the first place, even for Jython 2.1.  Have fixed.

>  * 3725.removal needs a final carriage return

Fixed

>  * test_threadpool.py and threadpool.py have trailing whitespace

Fixed

>  * does serialport.py really not have any test cases? I ask because it doesn't have a test case name at the top, but grepping around I didn't see anything. Too bad.

Oh well.  :)  Maybe I will look at those tickets some time (perhaps not near) in the future.

>  * the for loop at the bottom of test_threadpool.py for setting skip uses DeprecationTestCase when the test class's name is ThreadSafeListDeprecationTestCase. This [doesn't appear to have actually broken anything](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/jython-platformdetection-remove-3725), so is the surrounding interface check useless?

Huh, you are correct.  I had originally named it DeprecationTestCase, and then changed it to be more clear, but apparently none of that stuff mattered.  So I took out the interface check, and moved the "else" imports up top.

>  * pyflakes warnings: test_threadpool.py imports Version but doesn't use it. threadpool.py imports runtime but doesn't use it. serialport.py has a bunch of unused imports. jelly.py imports runtime but doesn't use it.

threadpool.py also imports runtime but doesn't use it.  Removed the unused imports from all mentioned files.

>  * grepping through the source, I see a couple of other places where something about Java is special-cased, although I didn't check if they fall under the umbrella of this ticket: setup.py, twisted/python/runtime.py, and twisted/test/test_pb.py.

Regarding setup.py - extensions are a CPython thing, and I don't think they're needed in Java?  Anyway, I thought this ticket covered Twisted itself, rather than the installation/distribution/compilation, so I thought altering setup.py should be a different ticket.

The java-related stuff in twisted/python/runtime.py is regarding what platform Twisted is running on - it isn't special-casing anything exactly, it's just providing information.

Regarding twisted/test/test_pb.py: I originally interpreted this as not being covered by this ticket, because the test seemed more like the inverse of support for jython - it was explicitly not messing with jython's garbage collection, which I thought may have been intentional somehow because perhaps garbage collection would always be special in every version of jython.  But checking the <a href="http://www.jython.org/docs/library/gc.html">current jython docs regarding the garbage collector</a>, I see I'm probably wrong, so I've removed the java stuff in test_pb.py.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_10"></a>|@cyli commented|
|-|-|

(In [28767]) Fixed whitespace/extra imports/carriage returns as per review, and removed some extra Java special case code + broken test code 

refs [#3725](https://github.com/twisted/twisted/issues/3725)
-This line, and those below, will be ignored--

M    twisted/test/test_pb.py
M    twisted/test/test_threadpool.py
M    twisted/python/threadpool.py
M    twisted/topfiles/3725.removal
M    twisted/internet/serialport.py
M    twisted/spread/jelly.py

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)|@cyli removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_13"></a>|@jesstess set owner to @cyli|
|-|-|

test_pb.py needs a copyright bump, and test_threadpool.py could use a module-level docstring. Other than that, this branch looks great and ready to merge! Thanks for working on this, cyli.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_14"></a>|@cyli commented|
|-|-|

Replying to [jesstess](#note_13):
> test_pb.py needs a copyright bump, and test_threadpool.py could use a module-level docstring. Other than that, this branch looks great and ready to merge! Thanks for working on this, cyli.

Does that mean I should make the changes and kick it back into review, or change and merge?  Thanks!  :)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_15"></a>|@thijstriemstra commented|
|-|-|

Replying to [cyli](#note_14):
> Replying to [jesstess](#note_13):
> > test_pb.py needs a copyright bump, and test_threadpool.py could use a module-level docstring. Other than that, this branch looks great and ready to merge! Thanks for working on this, cyli.
> 
> Does that mean I should make the changes and kick it back into review, or change and merge?  Thanks!  :)

Can you put it back up for review and also:

 - replace the 'Twisted 10.1.0' reference with 'Twisted 10.1'
 - you mention that in 3725.removal that 'all internal Jython platform detection removed' but there's still a check in setup.py (that needs to stay) which skips building extensions on jython, so technically not all platform detection has been removed imo. Can you rephrase that and do something like 'in Twisted core' or 'across all projects' instead?

Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_16"></a>|@cyli commented|
|-|-|

(In [28859]) Change deprecation version to 10.1 instead of 10.0, copyright bump, and other minor branch description fixes as per review.

refs [#3725](https://github.com/twisted/twisted/issues/3725)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_17"></a>|@cyli removed owner|
|-|-|

Apologies for taking so long for such minor changes - have been both extremely busy and lazy.  Made the following modifications:

>  test_pb.py needs a copyright bump
>  test_threadpool.py could use a module-level docstring
>  - replace the 'Twisted 10.1.0' reference with 'Twisted 10.1'
>  - you mention that in 3725.removal that 'all internal Jython platform detection removed' but there's still a check in setup.py (that needs to stay) which skips building extensions on jython, so technically not all platform detection has been removed imo. Can you rephrase that and do something like 'in Twisted core' or 'across all projects' instead?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph set owner to @cyli|
|-|-|

Replying to [cyli](#note_17):

> >  - replace the 'Twisted 10.1.0' reference with 'Twisted 10.1'

I don't understand why thijs asked for this; the standard is currently to have all of major/minor/micro versions listed in the deprecation warning so that the version is precisely identified.  In any case, `Version` requires a micro-version argument to its constructor, so lots of tests now fail in this branch with this traceback:
```
  File "twisted/python/threadpool.py", line 289, in ThreadSafeList
    Version("Twisted", 10, 1),
exceptions.TypeError: __init__() takes at least 5 arguments (4 given)
```
So please put the '.0' back.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)|@cyli removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph set owner to @cyli|
|-|-|

Unfortunately the Jython buildbot is offline right now, so I can't see if this branch makes any difference, but the other [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/jython-platformdetection-remove-3725) are encouraging, so please go ahead and merge!

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_22"></a>|@cyli set status to `closed`|
|-|-|

(In [28981]) Merge jython-platformdetection-remove-3725: jython platform detection removal


Author: cyli 
Reviewer: jesstess, thijs, glyph 
Fixes: [#3725](https://github.com/twisted/twisted/issues/3725)


twisted.python.threadpool.ThreadSafeList is deprecated and
Jython platform detection in Twisted core removed

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

`twisted.internet.serialport` used to export these names:

PARITY_NONE, PARITY_EVEN, PARITY_ODD, STOPBITS_ONE, STOPBITS_TWO, FIVEBITS, SIXBITS, SEVENBITS, EIGHTBITS

Abstractly, the removal is an incompatible change.

It appears as though at least one project is using them (fizzjik).  So it's also a real-world problem.  I think this should be reverted and the incompatible removals dropped.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_24"></a>|@cyli set status to `reopened`|
|-|-|

(In [28984]) Revert r28981: Revert jython platform detection removal

twisted.internet.serialport used to export these names:

PARITY_NONE, PARITY_EVEN, PARITY_ODD, STOPBITS_ONE, STOPBITS_TWO, FIVEBITS, SIXBITS, SEVENBITS, EIGHTBITS

Abstractly, the removal is an incompatible change.

It appears as though at least one project is using them (fizzjik). So it's also a real-world problem. I think this should be reverted and the incompatible removals dropped.

Reopens: [#3725](https://github.com/twisted/twisted/issues/3725)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_25"></a>|@cyli commented|
|-|-|

(In [28986]) Branching to 'jython-platformdetection-remove-3725-2'

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_26"></a>|@cyli removed owner<br>@cyli set status to `new`|
|-|-|

Reverted.  I've also added an `__all__` to explicitly export these variables so that the next person who runs pyflakes doesn't remove the imports.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_27"></a>|@cyli set status to `reopened`|
|-|-|

(In [28984]) Revert r28981: Revert jython platform detection removal

twisted.internet.serialport used to export these names:

PARITY_NONE, PARITY_EVEN, PARITY_ODD, STOPBITS_ONE, STOPBITS_TWO, FIVEBITS, SIXBITS, SEVENBITS, EIGHTBITS

Abstractly, the removal is an incompatible change.

It appears as though at least one project is using them (fizzjik). So it's also a real-world problem. I think this should be reverted and the incompatible removals dropped.

Reopens: [#3725](https://github.com/twisted/twisted/issues/3725)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_28"></a>|@cyli commented|
|-|-|

(In [28984]) Revert r28981: Revert jython platform detection removal

twisted.internet.serialport used to export these names:

PARITY_NONE, PARITY_EVEN, PARITY_ODD, STOPBITS_ONE, STOPBITS_TWO, FIVEBITS, SIXBITS, SEVENBITS, EIGHTBITS

Abstractly, the removal is an incompatible change.

It appears as though at least one project is using them (fizzjik). So it's also a real-world problem. I think this should be reverted and the incompatible removals dropped.

Reopens: [#3725](https://github.com/twisted/twisted/issues/3725)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun set owner to @cyli<br>@exarkun set status to `new`|
|-|-|

Looks good, please merge, thanks!

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_30"></a>|@cyli set status to `closed`|
|-|-|

(In [28995]) Merge jython-platformdetection-remove-3725-2: removing jython platform detection 

Author: cyli 
Reviewer: jesstess, thijs, glyph, exarkun 
Fixes: [#3725](https://github.com/twisted/twisted/issues/3725)

twisted.python.threadpool.ThreadSafeList is deprecated and Jython platform detection in Twisted core removed

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
